src/hooks/__tests__/use-load-data-pages.test.ts

```diff
@@ -0,0 +1,318 @@
+      viewService = {
+        gridColumnStartIndex: 20,
+        gridWidth: 10,
+        gridRowStartIndex: 20,
+        gridHeight: 10,
+      };
+      expect(shouldFetchAdditionalData(qLastExpandedPos, viewService)).toBe(false);
+    });
+  });
+
+  describe("isMissingLayoutData", () => {
+    test("should return true in case of new page and while qTop is falling behind current page", () => {
+      pageInfo = {
+        ...pageInfo,
+        currentPage: 5,
+      };
+      expect(isMissingLayoutData(layoutService, pageInfo)).toBe(true);
+    });
+
+    test("should return true if we are missing data in x axis (columns)", () => {
+      layoutService = {
+        layout: {
+          qHyperCube: {
+            ...layoutService.layout.qHyperCube,
+            qPivotDataPages: [{ qArea: { qTop: 0, qHeight: 25, qWidth: 100, qLeft: 0 } }],
+          },
+        },
+      } as LayoutService;
+      expect(isMissingLayoutData(layoutService, pageInfo)).toBe(true);
+    });
+
+    test("should return true if we are missing data in y axis (rows)", () => {
+      layoutService = {
+        layout: {
+          qHyperCube: {
+            ...layoutService.layout.qHyperCube,
+            qPivotDataPages: [{ qArea: { qTop: 0, qHeight: 100, qWidth: 25, qLeft: 0 } }],
+          },
+        },
+      } as LayoutService;
+      expect(isMissingLayoutData(layoutService, pageInfo)).toBe(true);
+    });
+
+    test("should return false if both axis are fulfilled with data", () => {
+      layoutService = {
+        layout: {
+          qHyperCube: {
+            ...layoutService.layout.qHyperCube,
+            qPivotDataPages: [{ qArea: { qTop: 0, qHeight: 100, qWidth: 100, qLeft: 0 } }],
+          },
+        },
+      } as LayoutService;
+      expect(isMissingLayoutData(layoutService, pageInfo)).toBe(false);
+    });
+
+    test("should return fallback to default `qArea` if it is not provided (an enforced fetch trigger basically)", () => {
+      layoutService = {
+        layout: {
+          qHyperCube: {
+            ...layoutService.layout.qHyperCube,
+            qPivotDataPages: [{}],
+          },
+        },
+      } as LayoutService;
+      expect(isMissingLayoutData(layoutService, pageInfo)).toBe(true);
+    });
+  });
+
+  describe("useLoadDataPages", () => {
+    let getHyperCubePivotDataMock: jest.MockedFunction<() => Promise<EngineAPI.INxPivotPage[]>>;
+
+    const renderer = () =>
+      renderHook(() => useLoadDataPages({ model, layoutService, viewService, pageInfo })).result.current;
+
+    beforeEach(() => {
+      getHyperCubePivotDataMock = jest.fn();
+      model = {
+        getHyperCubePivotData: getHyperCubePivotDataMock,
+      } as unknown as Model;
+    });
+
+    test("should set `qPivotDataPages` from layout if there was any", () => {
+      renderer();
+      expect(useStateSetterMock).toHaveBeenCalledTimes(1);
+      expect(useStateSetterMock).toHaveBeenCalledWith(layoutService.layout.qHyperCube.qPivotDataPages);
```

Comment:
Can you not verify the return value of the hook instead of looking at the mocks?

Reply:
not actually, 
Here the challenge is `usePromise`!

we are **setting a state** within it which means **another re-render** of hook -> which causes an **infinite loop** in renderHook and jest kills it with failure error!

<img width="1277" alt="Screenshot 2023-05-29 at 16 26 03" src="https://github.com/qlik-oss/sn-pivot-table/assets/29652890/97204750-0041-43fb-b06e-d7627a4588c2">


So in my solution, I'm just mocking state and checking if the state setter has been called with proper argument, **relying on the guarantee that how `useState` works** makes me sure that if state updates with proper arguments, then I will have a proper output from hook.

Ofcourse better idea is welcome here!

Reply:
Do really understand why it would do that but I'm fine with the solution then 👍